python/lvmdrp/utils/metadata.py

```diff
@@ -0,0 +1,828 @@
+    exptime : float, optional
+        exposure time of the target frames, by default None
+    neon : bool, optional
+        whether is Neon lamp on or not, by default None
+    hgne : bool, optional
+        whether is HGNE lamp on or not, by default None
+    krypton : bool, optional
+        whether is Krypton lamp on or not, by default None
+    xenon : bool, optional
+        whether is Xenon lamp on or not, by default None
+    argon : bool, optional
+        whether is Argon lamp on or not, by default None
+    ldls : bool, optional
+        whether is LDLS lamp on or not, by default None
+    quartz : bool, optional
+        whether is Quartz lamp on or not, by default None
+    quality : int, optional
+        bitmask representing quality of the recution, by default None
+    stage : int, optional
+        bitmask representing stage of the reduction, by default None
+    status : int, optional
+        bitmask representing status of the reduction, by default None
+    observatory : str, optional
+        name of the observatory, by default 'lco'
+
+    Returns
+    -------
+    pandas.DataFrame
+        the grouped metadata filtered following the given criteria
+    """
+    # extract raw frame metadata
+    store = _load_store(observatory=observatory)
+    dataset = store["raw"]
+
+    # default output
+    default_output = pd.DataFrame(columns=list(zip(*MASTER_METADATA_COLUMNS))[0])
+
+    # extract MJD if given, else extract all MJDs
+    if mjd is not None:
+        try:
+            metadata = pd.DataFrame(dataset[str(mjd)][()])
+        except KeyError:
+            return default_output
+    else:
+        metadata = []
+        for mjd in dataset.keys():
+            metadata.append(dataset[str(mjd)][()])
+
+        if len(metadata) > 0:
+            metadata = pd.DataFrame(np.concatenate(metadata, axis=0))
+        else:
+            return default_output
+
+    # close store
+    store.file.close()
+
+    # convert bytes to literal strings
+    metadata = _decode_string(metadata)
+
+    logger.info(f"found {len(metadata)} frames in store")
+
+    # filter by exposure number, spectrograph and/or camera
+    metadata = _filter_metadata(
+        metadata=metadata,
+        mjd=mjd,
+        imagetyp=imagetyp,
+        camera=camera,
+        exptime=exptime,
+        neon=neon,
+        hgne=hgne,
+        krypton=krypton,
+        xenon=xenon,
+        argon=argon,
+        ldls=ldls,
+        quartz=quartz,
+        quality=quality,
+        stage=stage,
+        status=status,
+    )
+    logger.info(f"final number of frames after filtering {len(metadata)}")
+
+    logger.info("grouping analogs")
+    metadata_groups = metadata.groupby(["imagetyp", "camera", "exptime"])
+
+    logger.info(f"found {len(metadata_groups)} groups of analogs:")
+    analogs = []
+    for g in metadata_groups.groups:
+        logger.info(g)
+        analogs.append(metadata_groups.get_group(g))
+    return analogs
+
+
+def match_master_metadata(
```

Comment:
Does this only return the relevant master biases and darks to use for a reduction?  Or it does also provide the closest matching reduced arc and flat to use for reducing a science frame?

Reply:
what happens if there are no matching relevant calibration frames?

Reply:
yes, this function will return relevant master calibration frames: bias, dark, fiberflat and arcs. The general structure of the output is a dictionary with calibration frame `imagetyp`s for keys and values being the relevant master calibration metadata (as a pandas series).

If no calibration frames are not found, the values will be `None`.

Reply:
Ok, great.  I will test this out.  So in principle reducing 60041 will find the right cals from 60039-40?